### PR TITLE
suppress all Autograd keys in AutoNonVariableTypeMode

### PR DESCRIPTION
**Pull request by bhosmer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42658 Include/ExcludeDispatchKeySetGuard API
* #42619 update DispatchKey::toString()
* **#42610 suppress all Autograd keys in AutoNonVariableTypeMode**

Fix for https://github.com/pytorch/pytorch/issues/42609: `AutoNonVariableTypeMode` should suppress all autograd dispatch keys, not just `Autograd` (e.g. `XLAPreAutograd`, `PrivateUse<N>_PreAutograd`)

Differential Revision: [D22963408](https://our.internmc.facebook.com/intern/diff/D22963408)

### CODE REVIEW

**Comment by ezyang** [comment]
You should also do the reserved use private autograd keys too. (Also, probably should just code the RAII guard, so you don't have to keep banging on TLS separately for each ctor/dtor)

**Comment by bhosmer** [comment]
@ezyang updated

File: aten/src/ATen/core/LegacyTypeDispatch.h [hunk 1]
```diff
@@ -43,15 +43,26 @@ namespace at {
 // trace).  To unify the two, we would first have to move profiling and tracing
 // out of VariableType.
 
+// TODO centralize this as part of dispatch key revamp
```
Comment (ezyang) [inline aten/src/ATen/core/LegacyTypeDispatch.h:46]: YUPP. Probably somewhere in like c10 headers in the end.

File: c10/core/impl/LocalDispatchKeySet.h [hunk 2]
```diff
@@ -93,6 +93,21 @@ class C10_API ExcludeDispatchKeyGuard {
   bool prev_state_;
 };
 
+class C10_API ExcludeDispatchKeySetGuard {
```
Comment (ezyang) [inline c10/core/impl/LocalDispatchKeySet.h:96]: Maybe we can just get rid of ExcludeDispatchKeyGuard now, only support sets.

**Review by ezyang — APPROVED** [review]

File: c10/core/impl/LocalDispatchKeySet.h (hunk 2 above)
Comment (bhosmer) [reply c10/core/impl/LocalDispatchKeySet.h:80]: Went ahead and did this, for Include as well.

Comment (bhosmer) [reply c10/core/impl/LocalDispatchKeySet.h:96]: #42658

**Review by ailzhang — APPROVED** [review]

**Pull request closed without merging**